### PR TITLE
Target file handling

### PR DESCRIPTION
**Pull request by cstamas**
Original issue: Provisio was silently overwriting conflicting target files.

Changes:
* this continues on change 3622ba8f2e0a308e703abac5bb0df531728e0219
* remove commons-lang3 (inline that single method)
* introduce ProvisioVariables helper class w/ rudimentary "config" support
* introduce config: allowTargetOverwrite (def false) - this is breaking change and will break builds that already have conflicting files
* introduce config: fallbackTargetFileNameMode - this reverts to "normal" state, using resolved file name, but keeps @wendigo change of "GA".
* etc

### CODE REVIEW

**Comment by cstamas** [comment]
@gnodet

File: provisio-core/src/main/java/ca/vanzyl/provisio/ProvisioVariables.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (C) 2015-2024 Jason van Zyl
```
Comment (gnodet) [inline provisio-core/src/main/java/ca/vanzyl/provisio/ProvisioVariables.java:2]: You should change or remove that line, as Jason is not the author afaik.

File: provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java [hunk 2]
```diff
@@ -41,20 +46,33 @@ public WriteToDiskAction(ProvisioArtifact artifact, File outputDirectory) {
...
 
-    public void copy(File source, File target) {
+    private void write(ProvisioningContext context, ProvisioArtifact source, String targetPath) {
+        File target = new File(outputDirectory, targetPath);
         try {
-            if (!target.getParentFile().exists()) {
-                target.getParentFile().mkdirs();
+            Files.createDirectories(target.getParentFile().toPath());
+            if (ProvisioVariables.allowTargetOverwrite(context)) {
+                if (Files.isRegularFile(target.toPath())) {
```
Comment (gnodet) [inline provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java:62]: What happens if you rebuild without having cleaned ?

Comment (cstamas) [reply provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java:62]: Good question... Should I maybe use that change detecting stream from p-u?

Comment (gnodet) [reply provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java:62]: That would be very nice ! Ideally, the assemblies would not be rebuilt if nothing changed, but that may be a different thing.

Comment (wendigo) [reply provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java]: Woah, that would be a killer feature

File: provisio-core/src/main/java/ca/vanzyl/provisio/ProvisioVariables.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+     */
+    public static final String GA_SEPARATOR = VARIABLE_PREFIX + "gaSeparator";
+
+    /**
+     * The maximum filename length to allow when in mode {@link FallBackTargetFileNameMode#GA},
+     * defaults to 64.
+     */
+    public static final String GA_MAX_FILE_NAME_LENGTH = VARIABLE_PREFIX + "maxFileNameLength";
+
+    //
```
Comment (wendigo) [inline provisio-core/src/main/java/ca/vanzyl/provisio/ProvisioVariables.java:70]: redundant comment

File: provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java [hunk 4]
```diff
@@ -41,20 +46,33 @@ public WriteToDiskAction(ProvisioArtifact artifact, File outputDirectory) {
...
+                    logger.warn("Conflict: artifact {} overwrote existing file {}", artifact, targetPath);
+                }
+                Files.copy(artifact.getFile().toPath(), target.toPath(), StandardCopyOption.REPLACE_EXISTING);
+            } else {
+                if (Files.isRegularFile(target.toPath())) {
+                    throw new ProvisioningException(
+                            "Conflict: artifact " + artifact + " would overwrite existing file: " + targetPath);
+                }
+                logger.debug("Copying {} -> {}", artifact, targetPath);
+                Files.copy(artifact.getFile().toPath(), target.toPath());
```
Comment (wendigo) [inline provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java]: can we just not copy if the target file exists and is exactly the same (name, size)?

**Review by wendigo — APPROVED** [review]

File: provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java [hunk 5]
```diff
@@ -41,20 +49,37 @@ public WriteToDiskAction(ProvisioArtifact artifact, File outputDirectory) {
...
+            Files.createDirectories(target.getParentFile().toPath());
+            if (Files.exists(target.toPath())) {
+                if (ProvisioVariables.allowTargetOverwrite(context)) {
+                    logger.warn("Conflict: artifact {} overwrites existing file {}", artifact, targetPath);
+                } else {
+                    throw new ProvisioningException(
+                            "Conflict: artifact " + artifact + " would overwrite existing file: " + targetPath);
+                }
+            }
+            try (CachingOutputStream outputStream = new CachingOutputStream(target.toPath(), BUFFER_SIZE);
```
Comment (cstamas) [inline provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java:72]: Use explicit buffer size is not needed, as below transferTo will again use default (8k) buffer.

File: provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java [hunk 6]
```diff
@@ -41,20 +49,37 @@ public WriteToDiskAction(ProvisioArtifact artifact, File outputDirectory) {
...
+                    BufferedInputStream inputStream =
+                            new BufferedInputStream(new FileInputStream(artifact.getFile()), BUFFER_SIZE)) {
+                inputStream.transferTo(outputStream);
+                outputStream.close();
```
Comment (wendigo) [inline provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java:76]: outputStream is closed by the try-with-resource

Comment (cstamas) [reply provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java:76]: True, but calling it multiple times is no-op. The point here is to flush/close it and then check "is it changed". If changed, apply timestamp

Comment (wendigo) [reply provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java:76]: Oh, it's a plexus thing :)

**Review by wendigo — APPROVED** [review]
LGTM

File: provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java (hunk 4 above)
Comment (gnodet) [reply provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java]: I don't think that's really safe.  The underlying cache stream is quite fast and the big benefit is that it won't write anything if there's no change at all.

**Comment by cstamas** [comment]
Released proviso-archiver 1.0.1 (fixes zipSlip and UnArchiver uses p-u caching stream...) will wait for central sync and update more in this PR

**Comment by cstamas** [comment]
@gnodet @wendigo pushed my latest changes, please review again

IF ok, @wendigo pls release (and then don't forget to fix Trino build as this PR reverts name mapping to original A but will fail if overwrite happens).

File: provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java (hunk 2 above)
Comment (cstamas) [reply provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java]: problem is that provision is "usually" open up something, mangle, archive up something. UnArchiver now will not overwrite if files present, nor will WriteToDiskAction, but last element is Archiver (that am unsure how could we fix)

**Comment by cstamas** [comment]
Tested this PR w/ mvnd build (where issue was detected), is :heavy_check_mark:

**Review by gnodet — APPROVED** [review]

**Comment by wendigo** [comment]
@cstamas is it faster on consecutive run?

**Comment by cstamas** [comment]
@wendigo https://gist.github.com/cstamas/865f8605c9f87f511099c79b659c5f73

Small speedup, but there is a problem: proviso now reports conflicts w/o clean, as @gnodet said

**Comment by wendigo** [comment]
@cstamas if we switch conflict detection to false, this will work without clean, right?

**Comment by wendigo** [comment]
so we need to detect whether in the artifact set, we have two artifacts that resolve to the same filename, rather then checking for conflict during write, correct?

**Comment by cstamas** [comment]
yes

**Comment by cstamas** [comment]
Condt in https://github.com/jvanzyl/provisio/pull/118

**Pull request merged**